syntax = "proto3";
package skywalker;

enum MessageType {
  PROPOSER_SEND_NEW_VALUE = 0;
  PREPARE = 1;
  PREPARE_REPLY = 2;
  ACCEPT = 3;
  ACCEPT_REPLY = 4;
  NEW_CHOSEN_VALUE = 5;
  LEARNER_ASK_FOR_LEARN = 6;
  LEARNER_SEND_LEARNED_VALUE = 7;
  LEARNER_SEND_NOW_INSTANCE_ID = 8;
  LEARNER_COMFIRM_ASK_FOR_LEARN = 9;
}

message PaxosMessage {
  MessageType type = 1;
  uint64 node_id = 2;
  uint64 instance_id = 3;
  uint64 proposal_id = 4;
  uint64 proposal_node_id = 5;
  string value = 6;
  uint64 pre_accepted_id = 7;
  uint64 pre_accepted_node_id = 8;
  uint64 reject_for_promised_id = 9;
  uint64 now_instance_id = 10;
}

message CheckPointMessage {
  MessageType type = 1;
}

enum ContentType {
  PAXOS_MESSAGE = 0;
  CHECKPOINT_MESSAGE = 1;
}

message Content {
  ContentType type = 1;
  uint32 group_id = 2;
  uint32 version = 3;
  PaxosMessage paxos_msg = 4;
  CheckPointMessage checkpoint_msg = 5;
}

message AcceptorState {
  uint64 instance_id = 1;
  uint64 promised_id = 2;
  uint64 promised_node_id = 3;
  uint64 accepted_id = 4;
  uint64 accepted_node_id = 5;
  string accepted_value = 6;
}

message SystemVariables {
  repeated uint64 membership = 1;
}
